addons/tls13-middlebox-serverhello/bootstrap.js

```diff
@@ -0,0 +1,291 @@
+      }
+    });
+  });
+}
+
+// returns true if there is at least one non-builtin root certificate is installed
+async function isNonBuiltInRootCertInstalled() {
+  let certs = nsIX509CertListToArray(certDB.getCerts());
+
+  for (let cert of certs) {
+    let chain = await getCertChain(cert, CERT_USAGE_SSL_CA);
+
+    if (chain !== null && chain.length === 1 && !chain[0].isBuiltInRoot) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+async function getInfo(xhr) {
+  let result = {};
+
+  try {
+    let channel = xhr.channel;
+
+    // this is the most important value based on which we can find out the problem
+    channel.QueryInterface(Ci.nsIRequest);
+    result.status = getFieldValue(channel, "status");
+
+    let securityInfo = getFieldValue(channel, "securityInfo");
+
+    if (securityInfo instanceof Ci.nsITransportSecurityInfo) {
+      securityInfo.QueryInterface(Ci.nsITransportSecurityInfo);
+
+      // extract security state and error code by which we can identify the reasons the connection failed
+      result.securityState = getFieldValue(securityInfo, "securityState");
+      result.errorCode = getFieldValue(securityInfo, "errorCode");
+    }
+
+    if (securityInfo instanceof Ci.nsISSLStatusProvider) {
+      securityInfo.QueryInterface(Ci.nsISSLStatusProvider);
+      let sslStatus = getFieldValue(securityInfo, "SSLStatus");
+
+      if (sslStatus) {
+        sslStatus.QueryInterface(Ci.nsISSLStatus);
+
+        // in case cert verification failed, we need to extract the cert chain from failedCertChain attribute
+        // otherwise, we extract cert chain using certDB.asyncVerifyCertAtTime API
+        let chain = null;
+
+        if (getFieldValue(securityInfo, "failedCertChain")) {
+          chain = nsIX509CertListToArray(securityInfo.failedCertChain);
+        } else {
+          chain = await getCertChain(getFieldValue(sslStatus, "serverCert"), CERT_USAGE_SSL_SERVER);
+        }
+
+        // extracting sha256 fingerprint for the leaf cert in the chain
+        result.serverSha256Fingerprint = getFieldValue(chain[0], "sha256Fingerprint");
+
+        // check the root cert to see if it is builtin certificate
+        result.isBuiltInRoot = (chain !== null && chain.length > 0) ? getFieldValue(chain[chain.length - 1], "isBuiltInRoot") : null;
+
+        // record the tls version Firefox ended up negotiating
+        result.protocolVersion = getFieldValue(sslStatus, "protocolVersion");
+      }
+    }
+  } catch (ex) {
+    result.exception = ex.message;
+  }
+
+  return result;
+}
+
+function makeRequest(config) {
+  return new Promise((resolve, reject) => {
+    // put together the configuration and the info collected from the connection
+    async function reportResult(event, xhr) {
+      resolve(Object.assign({"event": event, "responseCode": xhr.status}, await getInfo(xhr)));
+      return true;
+    }
+
+    try {
+      let xhr = Cc["@mozilla.org/xmlextras/xmlhttprequest;1"].createInstance(Ci.nsIXMLHttpRequest);
```

Comment:
Not important but I think you can do the less-wordy `Cu.importGlobalProperties(["XMLHttpRequest"])` instead if you want.

Reply:
In the interest of laziness, I'm going to leave it as-is :)